corehq/tests/locks.py

```diff
@@ -0,0 +1,80 @@
+import logging
+from contextlib import contextmanager
+from threading import RLock
+from unittest.mock import patch
+
+import attr
+
+import dimagi.utils.couch as module
+
+log = logging.getLogger(__name__)
+
+
+def reentrant_redis_locks(test=None):
+    """Decorator/context manager to enable reentrant redis locks
+
+    This is useful for tests that do things like acquire a lock and then
+    fire off a celery task (which will usually be executed synchronously
+    in tests) before the lock is released.
+
+    Usage as decorator:
+
+        @reentrant_redis_locks
+        def test_something():
+            ...
+
+    Usage as context manager:
+
+        def test_something():
+            with reentrant_redis_locks():
+                ...
+    """
+    def get_reentrant_lock(key, **kw):
+        try:
+            lock = locks[key]
+        except KeyError:
+            lock = locks[key] = ReentrantTestLock(key, locks)
+        return lock
+
+    @contextmanager
+    def context():
+        with patch.object(module, "get_redis_client", client):
+            try:
+                yield
+            finally:
+                assert not locks, f"unreleased {locks.values()}"
+
+    locks = {}
+    client = TestRedisClient(get_reentrant_lock)
+    manager = context()
+    return manager if test is None else manager(test)
+
+
+@attr.s
+class TestRedisClient:
+    lock = attr.ib()
+
+    def __call__(self):
+        return self
```

Comment:
for my own education, what's going on here?

Reply:
This class (`TestRedisClient`) is instantiated on line 48 in this module. When the patched version of `get_redis_client` is called it invokes this `__call__` method, which must return an object that has a `lock()` method.